nutils/numeric.py

```diff
@@ -439,6 +439,61 @@ def binom(n, k):
     b *= i
   return a // b
 
+def full(shape, fill_value):
+  return numpy.lib.stride_tricks.as_strided(fill_value, shape, [0]*len(shape), writeable=False)
+
+def writeable(array):
+  '''Test if any base is writeable.'''
+
+  return array.flags.writeable or isinstance(array.base, numpy.ndarray) and writeable(array.base)
+
+def freeze(array):
+  '''Set all bases unwriteable.'''
+
+  while isinstance(array, numpy.ndarray):
+    array.flags.writeable = False
+    array = array.base
+
+def frozen(array, *args, **kwargs):
+  '''Set all bases unwriteable.'''
+
+  array = numpy.array(array, *args, **kwargs)
+  freeze(array)
+  return array
+
+def keepfrozen(func):
+  from weakref import ref
+  cache = {}
+
+  @functools.wraps(func)
+  def wrapper(*args):
+    key = []
+    for arg in args:
+      if not isinstance(arg, numpy.ndarray):
+        key.append((type(arg), arg))
+      elif not writeable(arg):
+        key.append(Py_buffer(arg))
+      else:
+        return func(*args)
+    key = tuple(key)
+    try:
+      v = cache[key]
+    except KeyError:
+      v = cache[key] = func(*args)
+      freeze(v)
+      pop = functools.partial(cache.pop, key)
+      for buf in key:
+        if isinstance(buf, Py_buffer):
+          # replace the buffer exporter by a weak reference of the lowest base,
+          # with a callback at destruction to remove the cache entry
+          obj = buf.obj
+          while isinstance(obj, numpy.ndarray) and obj.base is not None:
```

Comment:
Should it be `while isinstance(obj, (numpy.ndarray, numpy.lib.stride_tricks.DummyArray)) ...` as now cache for views created by `stride_tricks` will be poped earlier?

Reply:
I think you missed this one in the latest version.

Reply:
Indeed! Thanks.